### PR TITLE
Add stream support to rpc.

### PR DESCRIPTION
**Pull request by sbunce**
Add support for client and server streaming.

Example:
```
service Serv {
	rpc Foo(stream Bar) returns (stream Baz);
}
```

### CODE REVIEW

**Comment by sbunce** [comment]
David,
My goal is to use gotoc with grpc. Any pointers on what needs to get done are appreciated. Thanks!

Seth

**Comment by sbunce** [comment]
I thought of a problem with this CL. There may be a message called "stream". I'll think about how to handle this.
```
message stream {}

service Serv {
    rpc Foo(stream) returns (stream);
}
```

**Comment by sbunce** [comment]
I was able to resolve that ambiguity. It now handles the corner case where a message is called "stream".  I added a test for it. The code is more complicated but I can't think of a way to make it simpler.

```
message stream {}

service Foo {
rpc Bar(stream) returns (stream);
}
```

**Comment by dsymonds** [comment]
Yeah, this looks too complicated. There's other parts of the parser that effectively do a token lookahead. I think you can just parse the token, and if it is the literal "stream" then the next token is either the stream's message name or ")". Alternatively, read two tokens (there always has to be at least two left), and it'll either be "stream" and Msg, or Msg and ")".

**Comment by dsymonds** [comment]
Also, please squash the commits when you push a new version.

**Comment by sbunce** [comment]
I checked in some garbage by accident. Trying to fix now. Please ignore last commit.

**Comment by sbunce** [comment]
I fixed my bad git push. Sorry for that noise.

I took a look at other parts of the parser where you use p.back(). I refactored this change to use p.back() and a few conditions. I think it's simpler to understand now.

**Comment by dsymonds** [comment]
Sorry I've been slow. I'll try to get to it this weekend.

File: parser/parser_test.go [hunk 1]
```diff
@@ -331,7 +331,19 @@ var parseTests = []parseTest{
 	{
 		"SimpleService",
 		"service TestService {\n  rpc Foo(In) returns (Out);\n}\n message In{} message Out{}",
-		`service { name: "TestService" method { name:"Foo" input_type:".In" output_type:".Out" } }` +
+		`service { name: "TestService" method { name:"Foo" input_type:".In" output_type:".Out" client_streaming: false server_streaming: false } }` +
```
Comment (dsymonds) [inline parser/parser_test.go]: This shouldn't have changed. client_streaming/server_streaming should not have been set if they were false.

File: parser/parser_test.go [hunk 2]
```diff
@@ -331,7 +331,19 @@ var parseTests = []parseTest{
...
+			`message_type:{name:"In"} message_type:{name:"Out"}`,
+	},
+	{
+		"SimpleServiceWithMessageCalledStream",
```
Comment (dsymonds) [inline parser/parser_test.go]: See src/google/protobuf/compiler/parser_unittest.cc in the protobuf repo for the test cases that I've adapted. I think this just needs MethodsAndStreams ported over.

File: gendesc/gendesc.go [hunk 3]
```diff
@@ -275,9 +275,11 @@ func genService(srv *ast.Service) (*pb.ServiceDescriptorProto, error) {
 
 func genMethod(mth *ast.Method) (*pb.MethodDescriptorProto, error) {
 	mdp := &pb.MethodDescriptorProto{
-		Name:       proto.String(mth.Name),
-		InputType:  proto.String(qualifiedName(mth.InType)),
-		OutputType: proto.String(qualifiedName(mth.OutType)),
+		Name:            proto.String(mth.Name),
+		InputType:       proto.String(qualifiedName(mth.InType)),
+		OutputType:      proto.String(qualifiedName(mth.OutType)),
+		ClientStreaming: proto.Bool(mth.ClientStreaming),
```
Comment (dsymonds) [inline gendesc/gendesc.go]: set these conditionally.

File: parser/parser.go [hunk 4]
```diff
@@ -781,15 +781,29 @@ func (p *parser) readService(srv *ast.Service) *parseError {
 		if err := p.readToken("("); err != nil {
 			return err
 		}
-
 		tok = p.next()
 		if tok.err != nil {
 			return tok.err
 		}
-		mth.InTypeName = tok.value // TODO: validate
+		if tok.value == "stream" {
```
Comment (dsymonds) [inline parser/parser.go]: What about something more like this?

mth.InTypeName = tok.value // TODO: validate
if tok.value == "stream" {
  // If the next token isn't ")", this is a stream.
  tok = p.next()
  if tok.err != nil { return tok.err }
  if tok.value != ")" {
    mth.InTypeName = tok.value
    mth.ClientStreaming = true
  } else {
    p.back()
  }
}

File: parser/parser.go [hunk 5]
```diff
@@ -781,15 +781,29 @@ func (p *parser) readService(srv *ast.Service) *parseError {
...
+				mth.ClientStreaming = true
+				mth.InTypeName = tok.value
+			}
+		} else {
+			mth.InTypeName = tok.value
+		}
 		if err := p.readToken(")"); err != nil {
 			return err
 		}
+
```
Comment (dsymonds) [inline parser/parser.go]: please avoid unrelated changes; revert this.

File: gendesc/gendesc.go (hunk 3 above)
Comment (sbunce) [inline gendesc/gendesc.go]: Done.

File: parser/parser.go (hunk 4 above)
Comment (sbunce) [inline parser/parser.go]: Refactored both the In and Out stuff to be like this.

File: parser/parser.go (hunk 5 above)
Comment (sbunce) [inline parser/parser.go]: Done.

Also reverted a few other unrelated changes.

File: parser/parser_test.go (hunk 1 above)
Comment (sbunce) [inline parser/parser_test.go]: Done

File: parser/parser_test.go (hunk 2 above)
Comment (sbunce) [inline parser/parser_test.go]: My intention here is to test the special case where there's a message called "stream". I did this because my first version didn't handle this special case and it didn't seem obvious to me. This could just be me though. I took a look at parser_unittest.cc and it doesn't look like it tests for this special case.

Comment (dsymonds) [inline parser/parser_test.go]: Ah, I missed that. Right. Okay, keep this test. You're right that MethodsAndStreams looks pointless; I should have looked more closely.

**Comment by dsymonds** [comment]
Yes, please just squash and force push. The GH review system is still horrid. I only want to merge a single commit, and this change is small enough to re-review in total each time.

**Comment by sbunce** [comment]
I squashed and force pushed.

File: gendesc/gendesc.go [hunk 6]
```diff
@@ -279,6 +279,12 @@ func genMethod(mth *ast.Method) (*pb.MethodDescriptorProto, error) {
 		InputType:  proto.String(qualifiedName(mth.InType)),
 		OutputType: proto.String(qualifiedName(mth.OutType)),
 	}
+	if mth.ClientStreaming {
+		mdp.ClientStreaming = proto.Bool(mth.ClientStreaming)
```
Comment (dsymonds) [inline gendesc/gendesc.go]: just use `proto.Bool(true)` in these two places.

File: parser/parser.go [hunk 7]
```diff
@@ -787,6 +787,21 @@ func (p *parser) readService(srv *ast.Service) *parseError {
...
+			if tok.err != nil {
+				return tok.err
+			}
+			if tok.value != ")" {
+				mth.InTypeName = tok.value
+				mth.ClientStreaming = true
+			} else {
+				p.back()
+			}
+		} else {
```
Comment (dsymonds) [inline parser/parser.go]: this `else` branch shouldn't be needed now.

File: parser/parser.go [hunk 8]
```diff
@@ -802,6 +817,21 @@ func (p *parser) readService(srv *ast.Service) *parseError {
...
+			if tok.err != nil {
+				return tok.err
+			}
+			if tok.value != ")" {
+				mth.OutTypeName = tok.value
+				mth.ServerStreaming = true
+			} else {
+				p.back()
+			}
+		} else {
```
Comment (dsymonds) [inline parser/parser.go]: ditto

File: parser/parser_test.go [hunk 9]
```diff
@@ -335,6 +335,18 @@ var parseTests = []parseTest{
 			`message_type:{name:"In"} message_type:{name:"Out"}`,
 	},
 	{
+		"SimpleServiceWithMessageCalledStream",
+		"message stream {}\nservice TestService {\n  rpc Foo(stream) returns (stream);\n}",
```
Comment (dsymonds) [inline parser/parser_test.go]: please put the message type at the end of the input, like SimpleService.

File: gendesc/gendesc.go (hunk 6 above)
Comment (sbunce) [inline gendesc/gendesc.go]: Ah, good point.

Done.

File: parser/parser.go (hunk 7 above)
Comment (sbunce) [inline parser/parser.go]: Thanks! I should have seen that.

File: parser/parser.go (hunk 8 above)
Comment (sbunce) [inline parser/parser.go]: Done

File: parser/parser_test.go (hunk 9 above)
Comment (sbunce) [inline parser/parser_test.go]: Done.

**Comment by sbunce** [comment]
Thanks for your patience on this review. I should have noticed that extraneous else branch.

I pushed an updated version.

**Comment by dsymonds** [comment]
Merged in 6e9e04d.

Thanks for your contribution!
